### ISSUE TITLE
MLP

### ISSUE THREAD

**Issue by MonZop** [MonZop/BioBlender#3]
Besides the facty that the surface is necessary to display the MLP (see Issue #1), there seems to be also another problem: apparently BB cannot find pyMLP.py (which is distributed with BB, in the Bin folder)

**Comment by zeffii** [comment]
I have a question about this
https://github.com/MonZop/BioBlender/blob/master/BioBlender2.py#L2078-L2093

It seems a little odd to have `linux` and `darin` execution paths (code flow) without an explicit windows perhaps a second elif should be explicit for windows.

**Comment by zeffii** [comment]
OK, i think I understand.

**Comment by MonZop** [comment]
That's good, because I am lost :(

**Comment by MonZop** [comment]
(from Issue #1 Surface)
 I do get the PyMOL generated surface.
No MLP visulization at all; not atomic and not integrated on the surface...

**Comment by MonZop** [comment]
![untitled](https://cloud.githubusercontent.com/assets/10231334/6061199/ee9f7f96-ad45-11e4-85df-f3c1db3c5697.png)
this is the first mapping in grey scale. Then it is rendered as a combined  scale : from dark, opaque, rough to white, shiny, smooth (hydrophilic to lipophilic)

![rendertest](https://cloud.githubusercontent.com/assets/10231334/6061303/d582b112-ad46-11e4-8314-e730c78395a0.png)
rendered (in this case only rough/smooth and matte/shiny, without the color gradient)

![atomicmlp](https://cloud.githubusercontent.com/assets/10231334/6078362/0f29c796-adf8-11e4-824b-f7beeae067ce.png)
And this is the Atomic MLP view.

**Comment by MonZop** [comment]
The logical steps are:
Use pyMLP to attribute to each atom its fi value (I was surprised to find the fi values in the main script)
Integrate the values usign the formula of your choice (Testa is default), and produce a file .dx containing the grid of values in the 3D space of the ,molecule.
For each vertex of the mesh, interpolate the local vaule in the grid, and attribute the value to the vertex, in a file .obj, with x,y,z and the resulting value in u
Now the file can be sent to Blender, that will:
transform the vertex u into a scale of greys (after normalizing the values to 0-1); and build a grey UV map;
use the  map for color; for adding bump and specular features to the mesh.
The atomic MLP is a simple attribution of grey to each atom, represented in the VanderWaals sphere.

**Comment by zeffii** [comment]
![current_version](https://cloud.githubusercontent.com/assets/619340/6091859/572a743c-aec0-11e4-917a-b7f53da676e7.gif)

I think this code does work, but the UI behaviour could be more intuitive. At present the user must click an atom, then press _atomic mlp_. Maybe i'm misunderstanding the issue?

**Comment by zeffii** [comment]
The colour gradient seems to be only partially implemented, or not imported correctly.
At the moment it produces this:
![image](https://cloud.githubusercontent.com/assets/619340/6092071/d05544bc-aecb-11e4-8107-51c6cd512083.png)

- MLP appears to be imported as a VertexColor Map. (mapping each vertex of each polygon to a color)
- this VertexColor map can be baked into a texture (that's what the Render MLP to surface appears to do)

**Comment by zeffii** [comment]
@MonZop 
I thought using more grid subdivisions of the MLP Surface would allow the gradient to be more gradual as it appears to be in your images. What are the UI settings you used (any specific MLP formula?).
Has this ever worked in the Addon version of BioBlender? If so I will be able to track down the differences introduced over time. Else simply code it fresh.

**Comment by MonZop** [comment]
![mlptest_bb2](https://cloud.githubusercontent.com/assets/10231334/6104490/25cf6bf6-b04e-11e4-9805-ce05e4d1be67.png)
You are right. Ithink I had not selected the molecule (any atom of a molecule).
When I do, i get the result above. also the render is with these colors

**Comment by zeffii** [comment]
This test scene is illuminated with a icosphere cage to give an even distribution in ambient light.

I think this shows in an amplified way the intended final effect?
![image](https://cloud.githubusercontent.com/assets/619340/6105106/f9a1bb1e-b053-11e4-8813-0bb97b1a4797.png)

The triangles visible on the surface of the sphere are due to the high gloss in that area (practically mirror-esque), the pitted / cratered area is exaggerated and would be replaced by some other preferable texture.

**Comment by MonZop** [comment]
I don't understand, sorry

**Comment by zeffii** [comment]
![spec_bump](https://cloud.githubusercontent.com/assets/619340/6107022/e8ac5950-b066-11e4-8d9f-1e46612c71f6.gif)
It appears that the bump and specular maps are already generated, they seem correct even.

**Comment by MonZop** [comment]
are you using the Master_main_first_fixes ?

**Comment by zeffii** [comment]
yes Master_main_first_fixes, 

I manually added a small node material setup (not cycles..)
![image](https://cloud.githubusercontent.com/assets/619340/6107124/2b6b5c7c-b068-11e4-9f50-5ff6e122ac86.png)

It is easy to programmatically generate node setups for materials / textures, if we find a nice configuration. Nodes are a little bit more flexible than the original material system, and i'd like to make it compatible with Cycles which relies on nodes for the routing.

**Comment by MonZop** [comment]
I suppose you are using the very small peptide 1l2y. Ther distribution of the lipophilic potential on such a small surface is hard to assess. Could you test Calmodulin (File 01_CaM)? thanks
I will try to set up a node system like yours

**Comment by zeffii** [comment]
sure

**Comment by zeffii** [comment]
![zoom_shows_bump](https://cloud.githubusercontent.com/assets/619340/6107535/01056852-b06c-11e4-956a-fe5ea84d7de1.gif)
A zoom

**Comment by MonZop** [comment]
Looks OK (except the green spot, visible at the start). Can you see a more intense bump in the darker area?

**Comment by zeffii** [comment]
the green was the 'world orientation' marker (red, green, blue)

**Comment by MonZop** [comment]
Ah, OK. I thought it might be a mapping problem (normal inverted or something like)

**Comment by zeffii** [comment]
it appeats the bump pattern is created: 
![image](https://cloud.githubusercontent.com/assets/619340/6108159/246ddd06-b071-11e4-80c9-23f44bfe4ec4.png)
but what isn't happening automatically yet (or i am missing something) is the multiplication with the vertex color map to help elevate certain regions (enhance the bump) and smooth the rest

**Comment by MonZop** [comment]
![mlp mapping flow](https://cloud.githubusercontent.com/assets/10231334/6108474/e672fefc-b073-11e4-87e2-89cbdfd541b0.png)
This was the flow in the earlier versions of BioBlender. In theory, it should be still the same!

**Comment by MonZop** [comment]
Just a little explanation:
Calculation of MLP is done frame by frame, because as the protein changes conformation, it may hide or expose patches with different MLP values.
pyMLP includes an Atom Type Detection Library, which is used to assign a value (fi) to every atom in the pdb file. The Testa formula decays with a distance exponential and we added it to the original pyMLP, since we consider it a better function for MLP.
OBJcreator is a program that uses the mesh (created by pyMOL) and the grid of values (created by pyMLP), and atributes to each vertex of the mesh the value of MLP. This is typically between -3 and 1. The MLP value is stored in the obj file, in the V field (the U filed is left blank).
The file .obj is opened in Blender, where it is importe as image texture; the V values are converted to grey (between 0 and 1); and used in 3 steps:
direct color (grey scale); intensity of specular; intensity of noise which is then used for generating bumps.
Hope it is clear enough, in case it is not, just ask!

**Comment by zeffii** [comment]
ObjCreator I see, this does clarify the sudden presence of the vertex_colors. Another piece of the puzzle just fell into place.

**Comment by zeffii** [comment]
@MonZop Time permitting I will merge these explanations into https://github.com/MonZop/BioBlender/issues/19 for convenience

**Comment by zeffii** [comment]
@MonZop  
> Can you see a more intense bump in the darker area?

No sign of that. Furthermore, should the lighter area show more smoothing? All the right textures / vertex_color_maps are there, but they aren't being hooked up as in the diagram. Should be a minor issue to fix for both Blender Internal materials and Cycle materials (nodes)

**Comment by zeffii** [comment]
a little exaggerated 
![tensor](https://cloud.githubusercontent.com/assets/619340/6133548/f7432c6a-b15a-11e4-9061-678b0728263c.png)

**Comment by zeffii** [comment]
![tentacles](https://cloud.githubusercontent.com/assets/619340/6135129/55b8fbf2-b166-11e4-895f-b36ae08e3ed9.PNG)

**Comment by MonZop** [comment]
Seems OK. How did you get it?

**Comment by zeffii** [comment]
The 3 components are made by the existing code. 

- specular texture
- bump texture
- greyscale (vertex colour map)

The example image is the result of a material node setup which combines these elements, I would like to optimize the node setup first before presenting it as a realistic option.

By the way, in an ideal world, are these maps supposed to change per frame. These hydrophilic / lipophilic  areas would move around as a result of molecular motion?

**Comment by MonZop** [comment]
Yes!, this is exaclty the issue: both the surface and its features (MLP and EP) change frame by frame when the protein moves.
Several groups around theworld are working to optimize the calculations in order to allow real time renders. For relatively small systems (up to few thousands atoms) results have been obtained by distributing computations to GPUs or by parallelizing; of course the major laboratories have very larte dedicated machines. 
I am reviewing the literature to check if there are any mesh computing algorithm that are suitable for us, and might be faster and possibly avoid the need for external program (pyMOL).

**Comment by MonZop** [comment]
I just noticed that the image above is probably not what we want. 
This si due to the fact that pyMOL attributes colors to the mesh, related to the identity of the underlying atom; I think that this is the color 'transformed' accoridng to the MLP code. What we want to do is substitute those colors (i do not know how they are encoded in the wrl file) with the 'color' (single value) of the lipophilic potential, calculated with pyMLP.
If this is not clear I can try to explain better

**Comment by zeffii** [comment]
Ultimately if the lipophilic potential is one value per mesh vertex, then you get a grey scale or 3 identical values for rgb per vertex color. I'm pretty sure this is already the case.. but will read that section of the code again.

from my recollection the first import of the wrl writes results in a vertex_color_map, a greyscale map, subsequently doing the 'render MLP' performs different code some of which is commented out prior to my involvement. Will read this and get a better mental model. 

Feel free to elaborate.

**Comment by zeffii** [comment]
out of interest: https://github.com/MonZop/BioBlender/issues/3#issuecomment-73238861
that first image has a light patch in the lower part, that's not due to light is it? that's the vertex color map? Could you upload or email the resulting .blend with that file..so I can look at the vertex_color_map ?

**Comment by MonZop** [comment]
I htink that there is a missing step: the import of the 'lipophilic values' (colors), that should replace the colors provided by pyMOL. The vertex color map imported is not grey scale (same value for the 3 rgb), but real colors: indeed , if you render the mesh, it will show in green, red and blue (Carbon, Oxygen and Nitrogen, respectively).
I am just thinking of a possible different (and possibly faster) way to paint the MLP on the mesh:
each vertex of the mesh is close to one atom in the underlying structure. Each one of these atoms has a single MLP value, ranging from -3 to +1, and normalized from 0 to 1.  This can be seen as 'atomic MLP'
Is it possible to attribute that value to the vertex?

**Comment by zeffii** [comment]
yes, probably, proximity between a point and all other points is a relatively fast operation using a _k-dimensional tree_

**Comment by MonZop** [comment]
Yes, the bright spot is precisely indication of a very lipophilic patch in that part of the protein, not light.
For the file, I can send it tomorrow: it was obtained with BioBlender 0.6, the patched precompiled version, that (for reason that do not understand) does not work on this pc. But i can do it with the pc at home!

**Comment by MonZop** [comment]
Also consider that there is a quite short space of min-max distance between the vertex point and the closest atom

**Comment by MonZop** [comment]
(this conversation is running on two trakcs, i have no problem, but it might be confusing for potential onlookers)

**Comment by zeffii** [comment]
It is better to separate the issues, i agree. More granular approach to issue tracking is preferred, it allows sharp delineation.

**Comment by MonZop** [comment]
OK, I start a new issue: 'new MLP' , and move there the relevant parts
